File: src/07.greedy/01.arrayGreedy/09.621.ts
export function leastInterval(tasks: string[], n: number): number {
    const buckets = new Array(26).fill(0);
    for (let i = 0; i < tasks.length; i++) {
        buckets[tasks[i].charCodeAt(0) - "A".charCodeAt(0)]++;
    }

    buckets.sort((a, b) => a - b);

    let maxTimes = buckets[25];
    let maxCount = 1;

    for (let i = 25; i >= 1; i--) {
        if (buckets[i] === buckets[i - 1]) {
            maxCount++;
        } else {
            break;
        }
    }

    const res = (maxTimes - 1) * (n + 1) + maxCount;

    return Math.max(res, tasks.length);
}

File: src/07.greedy/01.arrayGreedy/09.621.go
package array_greedy

import "sort"

func LeastInterval(tasks []byte, n int) int {
	buckets := make([]int, 26)
	for _, v := range tasks {
		buckets[v-'A']++
	}

	sort.Ints(buckets)

	maxTimes := buckets[25]
	maxCount := 1

	for i := 25; i >= 1; i-- {
		if buckets[i] == buckets[i-1] {
			maxCount++
		} else {
			break
		}
	}

	res := (maxTimes-1)*(n+1) + maxCount

	if res > len(tasks) {
		return res
	}

	return len(tasks)
}
